src/Npgsql/NpgsqlConnector.FrontendMessages.cs

```diff
@@ -0,0 +1,420 @@
+            }
+
+            void Write(int len, StatementOrPortal statementOrPortal, string name)
+            {
+                WriteBuffer.WriteByte(FrontendMessageCode.Describe);
+                WriteBuffer.WriteInt32(len - 1);
+                WriteBuffer.WriteByte((byte)statementOrPortal);
+                WriteBuffer.WriteNullTerminatedString(name);
+            }
+        }
+
+        internal Task WriteSync(bool async)
+        {
+            if (WriteBuffer.WriteSpaceLeft < 5)
+                return FlushAndWrite(async);
+
+            Write();
+            return Task.CompletedTask;
+
+            async Task FlushAndWrite(bool async)
+            {
+                await Flush(async);
+                Debug.Assert(5 <= WriteBuffer.WriteSpaceLeft, $"Message of type {GetType().Name} has length 5 which is bigger than the buffer ({WriteBuffer.WriteSpaceLeft})");
+                Write();
+            }
+
+            void Write()
+            {
+                WriteBuffer.WriteByte(FrontendMessageCode.Sync);
+                WriteBuffer.WriteInt32(4);
+            }
+        }
+
+        internal Task WriteExecute(int maxRows, bool async)
+        {
+            // Note: non-empty portal currently not supported
+
+            if (WriteBuffer.WriteSpaceLeft < 10)
+                return FlushAndWrite(maxRows, async);
+
+            Write(maxRows);
+            return Task.CompletedTask;
+
+            async Task FlushAndWrite(int maxRows, bool async)
+            {
+                await Flush(async);
+                Debug.Assert(10 <= WriteBuffer.WriteSpaceLeft, $"Message of type {GetType().Name} has length 10 which is bigger than the buffer ({WriteBuffer.WriteSpaceLeft})");
+                Write(maxRows);
+            }
+
+            void Write(int maxRows)
+            {
+                WriteBuffer.WriteByte(FrontendMessageCode.Execute);
+                WriteBuffer.WriteInt32(9);
+                WriteBuffer.WriteByte(0);   // Portal is always an empty string
+                WriteBuffer.WriteInt32(maxRows);
+            }
+        }
+
+        internal async Task WriteParse(string sql, string statementName, List<NpgsqlParameter> inputParameters, bool async)
+        {
+            Debug.Assert(statementName != null && statementName.All(c => c < 128));
+
+            var queryByteLen = TextEncoding.GetByteCount(sql);
+            if (WriteBuffer.WriteSpaceLeft < 1 + 4 + statementName.Length + 1)
+                await Flush(async);
+
+            var messageLength =
+                1 +                         // Message code
+                4 +                         // Length
+                statementName.Length +
+                1 +                         // Null terminator
+                queryByteLen +
+                1 +                         // Null terminator
+                2 +                         // Number of parameters
+                inputParameters.Count * 4;
```

Comment:
Maybe `sizeof`s instead of numbers?

Reply:
You mean like `sizeof(byte)`, `sizeof(int)`? It's possible, but it would make everything considerably more verbose. But if you feel like it would improve readability I will do it.

Reply:
Yes, it will improve verbosity and make it easier to understand for what it is for (:

Reply:
OK, will do the change.